extensions/notebook/src/test/python/jupyterInstallation.test.ts

```diff
@@ -193,4 +201,92 @@ describe('Jupyter Server Installation', function () {
 		let commandStr = commandStub.args[0][0] as string;
 		should(commandStr.includes('"jupyter==1.0.0" "TestPkg2==4.5.6"')).be.true();
 	});
+
+	it('Get required packages test', async function() {
+		// Undefined argument
+		let packages = installation.getRequiredPackagesForKernel(undefined);
+		should(packages).not.be.undefined();
+		should(packages.length).be.equal(0);
+
+		// Invalid kernel name
+		packages = installation.getRequiredPackagesForKernel('NotARealKernel');
+		should(packages).not.be.undefined();
+		should(packages.length).be.equal(0);
+
+		// Python 3
+		let expectedPackages: PythonPkgDetails[] = [{
+			name: 'jupyter',
+			version: '1.0.0'
+		}];
+		packages = installation.getRequiredPackagesForKernel(python3DisplayName);
+		should(packages).be.deepEqual(expectedPackages);
+
+		// Powershell
+		expectedPackages.push({
+			name: 'powershell-kernel',
+			version: '0.1.3'
+		});
+		packages = installation.getRequiredPackagesForKernel(powershellDisplayName);
+		should(packages).be.deepEqual(expectedPackages);
+
+		// PySpark
+		expectedPackages = [{
+			name: 'jupyter',
+			version: '1.0.0'
+		}, {
+			name: 'sparkmagic',
+			version: '0.12.9'
+		}, {
+			name: 'pandas',
+			version: '0.24.2'
+		}, {
+			name: 'prose-codeaccelerator',
+			version: '1.3.0'
+		}];
+		packages = installation.getRequiredPackagesForKernel(pysparkDisplayName);
+		should(packages).be.deepEqual(expectedPackages);
+
+		packages = installation.getRequiredPackagesForKernel(sparkScalaDisplayName);
+		should(packages).be.deepEqual(expectedPackages);
+
+		packages = installation.getRequiredPackagesForKernel(sparkRDisplayName);
+		should(packages).be.deepEqual(expectedPackages);
+	});
+
+	it('Install python test', async function() {
+		// Should reject overwriting an existing python install if running on Windows and python is currently running.
+		if (process.platform === winPlatform) {
+			sinon.stub(utils, 'executeBufferedCommand').resolves('python.exe');
+
+			let installSettings: PythonInstallSettings = {
+				installPath: `${process.env['USERPROFILE']}\\ads-python`,
+				existingPython: false,
+				packages: []
+			};
+			await should(installation.startInstallProcess(false, installSettings)).be.rejected();
+
+			sinon.restore();
+		}
+
+		// Install required packages to existing python instance
+		let installSettings: PythonInstallSettings = {
+			installPath: `${process.env['USERPROFILE']}\\ads-python`,
+			existingPython: true,
+			packages: installation.getRequiredPackagesForKernel(python3DisplayName)
+		};
+		sinon.stub(utils, 'exists')
```

Comment:
Is there a reason you're only stubbing the 1st and 2nd call instead of all of them?

Reply:
Sometimes the exists calls each need to do different things. I'll add a fail case for this if we ever go past the second call. Half of this class is wrapped up in a single public method, so the mock-up flow gets pretty complicated.